src/meta/src/rpc/ddl_controller.rs

```diff
@@ -0,0 +1,590 @@
+                    .drop_sink(sink_id, internal_table_ids)
+                    .await?;
+                (version, vec![sink_id.into()])
+            }
+            StreamingJobId::Table(source_id, table_id) => {
+                self.drop_table_inner(source_id, table_id, internal_table_ids)
+                    .await?
+            }
+            StreamingJobId::Index(index_id) => {
+                let index_table_id = self.catalog_manager.get_index_table(index_id).await?;
+                let version = self
+                    .catalog_manager
+                    .drop_index(index_id, index_table_id)
+                    .await?;
+                (version, vec![index_table_id.into()])
+            }
+        };
+
+        self.stream_manager
+            .drop_streaming_jobs(streaming_job_ids)
+            .await;
+        Ok(version)
+    }
+
+    /// `prepare_stream_job` prepares a stream job and returns the context and table fragments.
+    async fn prepare_stream_job(
+        &self,
+        stream_job: &mut StreamingJob,
+        fragment_graph: StreamFragmentGraphProto,
+    ) -> MetaResult<(CreateStreamingJobContext, TableFragments)> {
+        // 1. Get the env for streaming jobs.
+        let env = StreamEnvironment::from_protobuf(fragment_graph.get_env().unwrap());
+
+        // 2. Build fragment graph.
+        let fragment_graph =
+            StreamFragmentGraph::new(fragment_graph, self.env.id_gen_manager_ref(), &*stream_job)
+                .await?;
+        let default_parallelism = fragment_graph.default_parallelism();
+        let internal_tables = fragment_graph.internal_tables();
+
+        // 3. Set the graph-related fields and freeze the `stream_job`.
+        stream_job.set_table_fragment_id(fragment_graph.table_fragment_id());
+        let dependent_relations = fragment_graph.dependent_relations().clone();
+        stream_job.set_dependent_relations(dependent_relations.clone());
+
+        let id = stream_job.id();
+        let stream_job = &*stream_job;
+
+        // 4. Mark current relation as "creating" and add reference count to dependent relations.
+        self.catalog_manager
+            .start_create_stream_job_procedure(stream_job)
+            .await?;
+
+        // 5. Resolve the upstream fragments, extend the fragment graph to a complete graph that
+        // contains all information needed for building the actor graph.
+        let upstream_mview_fragments = self
+            .fragment_manager
+            .get_upstream_mview_fragments(&dependent_relations)
+            .await;
+        let upstream_mview_actors = upstream_mview_fragments
+            .iter()
+            .map(|(&table_id, fragment)| {
+                (
+                    table_id,
+                    fragment.actors.iter().map(|a| a.actor_id).collect_vec(),
+                )
+            })
+            .collect();
+
+        let res: MetaResult<(CreateStreamingJobContext, TableFragments)> = try {
```

Comment:
Changed from previous version in `ddl_service`, should try here to call cancel when an error is caught.

Reply:
Can we simply use a big `try` at L220-232? We may ignore it in `cancel` if it fails even before `start_create_procedure`.

Reply:
I tried but it doesn't look feasible. If it fails in L220-232 before or inside `start_create_procedure`, it will wrongly unmark creating flags in `cancel` which is marked by others.